src/mesh/mesh.hpp

```diff
@@ -168,7 +170,15 @@ class Mesh {
   std::vector<LogicalLocation> GetLocList() const noexcept { return loclist; }
 
   void OutputMeshStructure(const int dim, const bool dump_mesh_structure = true);
-
+  
+    
+  
+  // Ordering here is important to prevent deallocation of pools before boundary 
+  // communication buffers 
+  using channel_key_t = std::tuple<int, int, std::string>;
+  std::map<int, buf_pool_t<Real>> pool_map;
+  std::map<channel_key_t, CommBuffer<buf_pool_t<Real>::owner_t>> boundary_comm_map;
```

Comment:
I'm curious if instead using an `unordered_map` here could show improvement on a real application. My understanding is there will be about `27 * [number of meshblocks on rank] * number of variables` entires in `boundary_comm_map`. `map` looks up in `O(log(N))` time whereas `unordered_map` on average looks up in `O(1)` time, and `log(27*100*50) = 5`. So one might get a ~5x speedup on the `IterateBoundaries` calls above. Not sure that this will ever matter, but it *might* help scaling efficiency at high processor count/high meshblock per processor count? Just a thought.

Reply:
I definitely agree that this should be changed. The problem was that `std::unordered_map` doesn't have a default hash that works with a tuple, so I just use `std::map` for now because it was quicker to get things working. I have never looked at what is required to create user defined hashes for `std::map` and if there is some good reason why there is not default available that works with a tuple.

Reply:
A simple fix might be to just replace tuple with a container struct.

Reply:
I couldn't get this to work with a struct. I just added some code to `utils` to hash tuples. I haven't really though about hash functions much, but what is in there seems to be a pretty standard function for combining hashes. Switching to `std::unordered_map` does seem to give a significant performance gain on the sparse advection test at least.